src/processor/fingerprint.cpp

```diff
@@ -0,0 +1,307 @@
+    string_hash_field(string_hash_field &&) = default;
+    string_hash_field &operator=(const string_hash_field &) = default;
+    string_hash_field &operator=(string_hash_field &&) = default;
+
+    std::size_t length() override { return 8; }
+    void operator()(string_buffer &output) override;
+
+    std::string_view value;
+};
+
+struct key_hash_field : field_generator {
+    explicit key_hash_field(const ddwaf_object &input) : value(input) {}
+    ~key_hash_field() override = default;
+    key_hash_field(const key_hash_field &) = default;
+    key_hash_field(key_hash_field &&) = default;
+    key_hash_field &operator=(const key_hash_field &) = default;
+    key_hash_field &operator=(key_hash_field &&) = default;
+
+    std::size_t length() override { return value.type == DDWAF_OBJ_MAP ? 8 : 0; }
+    void operator()(string_buffer &output) override;
+
+    ddwaf_object value;
+};
+
+template <typename T, typename... Rest>
+std::size_t generate_fragment_length(T &generator, Rest... rest)
+{
+    if constexpr (sizeof...(rest) > 0) {
+        return generator.length() + 1 + generate_fragment_length(rest...);
+    } else {
+        return generator.length();
+    }
+}
+
+template <typename T, typename... Rest>
+void generate_fragment_field(string_buffer &buffer, T &generator, Rest... rest)
+{
+    generator(buffer);
+    if constexpr (sizeof...(rest) > 0) {
+        buffer.append('-');
+        generate_fragment_field(buffer, rest...);
+    }
+}
+
+template <typename... Args>
+ddwaf_object generate_fragment(std::string_view header, Args... generators)
+{
+    std::size_t total_length = header.size() + 1 + generate_fragment_length(generators...);
+
+    string_buffer buffer{total_length};
+    buffer.append_lowercase(header);
+    buffer.append('-');
+
+    generate_fragment_field(buffer, generators...);
+
+    ddwaf_object res;
+    auto [ptr, size] = buffer.move();
+    ddwaf_object_stringl_nc(&res, ptr, size);
+
+    return res;
+}
+
+// Return true if the first argument is less than (i.e. is ordered before) the second
+bool str_casei_cmp(std::string_view left, std::string_view right)
+{
+    auto n = std::min(left.size(), right.size());
+    for (std::size_t i = 0; i < n; ++i) {
+        auto lc = ddwaf::tolower(left[i]);
+        auto rc = ddwaf::tolower(right[i]);
+        if (lc != rc) {
+            return lc < rc;
+        }
+    }
+    return left.size() <= right.size();
+}
+
+void normalize_string(std::string_view key, std::string &buffer, bool trailing_separator)
+{
+    buffer.clear();
+
+    if (buffer.capacity() < key.size()) {
+        // Add space for the extra comma, just in case
+        buffer.reserve(key.size() + 1);
+    }
+
+    for (auto c : key) {
+        if (c == ',') {
+            buffer.append(R"(\,)");
+        } else {
+            buffer.append(1, ddwaf::tolower(c));
+        }
+    }
+
+    if (trailing_separator) {
+        buffer.append(1, ',');
+    }
+}
```

Comment:
these functions operating byte by byte are good candidates for some optimizations in the future